Add slurm launcher service

Hyralc opened the pull request.

Hyralc committed:
Add usage of multiple launchers

Hyralc committed:
Add antares-launcher submodule

Hyralc committed:
change submodule name from antares-launcher to antares_launcher

Hyralc committed:
wip slurm launcher

Hyralc committed:
fix tests, change submodule location, add requirements

Hyralc committed:
feat: implement config and input data for slurm launcher

Hyralc committed:
wip: export flat

Hyralc committed:
add tests for slurm launcher

Hyralc committed:
Implement launcher run study and thread checking the study state

Hyralc committed:
fix mypy & black

Hyralc committed:
fix workflow
fix workflow
fix workflow for windows
fix workflow for windows
fix workflow for windows
Fix windows build

Signed-off-by: Paul Bui-Quang <paul.buiquang@rte-france.com>
fix black analyzing antareslauncher
fix black analyzing antareslauncher
fix black
fix black
fix black
fix black

Hyralc committed:
Fix threading problem

Hyralc committed:
Import working

Hyralc committed:
fix mypy

Hyralc committed:
fix test

Hyralc committed:
Fix tests after rebase

pl-buiquang committed:
Small changes

Signed-off-by: Paul Bui-Quang <paul.buiquang@rte-france.com>

pl-buiquang committed:
Fix jobid + change default level for paramiko lib logger

Signed-off-by: Paul Bui-Quang <paul.buiquang@rte-france.com>

pl-buiquang commented on antarest/launcher/business/slurm_launcher/slurm_launcher.py:
```diff
@@ -0,0 +1,241 @@
+
+    @staticmethod
+    def _delete_study(study_path: Path) -> None:
+        shutil.rmtree(study_path)
```
what the point of wrapping this into a static method ? it obfuscate the fact that it's a potentially dangerous operation.
it should at least check that the path is below the antares workspace directory.

pl-buiquang commented on antarest/launcher/business/slurm_launcher/slurm_launcher.py:
```diff
@@ -0,0 +1,241 @@
+        self.check_state = False
+        self.thread = None
+
+    def add_callback(
```
rename add_statusupdate_callback for more clarity

pl-buiquang commented on antarest/launcher/business/slurm_launcher/slurm_launcher.py:
```diff
@@ -0,0 +1,241 @@
+    def _delete_study(study_path: Path) -> None:
+        shutil.rmtree(study_path)
+
+    def _callback(
```
could rename "update_job_status"

pl-buiquang commented on antarest/launcher/business/slurm_launcher/slurm_launcher.py:
```diff
@@ -0,0 +1,241 @@
+        for path in glob.glob("**", recursive=True):
+            zipf.write(path, path)
+        zipf.close()
+        os.chdir(current_dir)
```
it would be better with a pure python version (https://docs.python.org/3/library/pathlib.html#module-pathlib) not relying on os.chdir and glob.glob

pl-buiquang commented on antarest/launcher/business/slurm_launcher/slurm_launcher.py:
```diff
@@ -0,0 +1,241 @@
+                    groups=[
+                        JWTGroup(id="admin", name="admin", role=RoleType.ADMIN)
+                    ],
+                )
```
refacto this and put a global variable representing "ADMIN" in common/jwt

pl-buiquang commented on antarest/launcher/service.py:
```diff
@@ -35,32 +34,44 @@ def __init__(
         study_version = study_info["antares"]["version"]
-        study_path = self.storage_service.get_study_path(study_uuid, params)
-        job_uuid: UUID = self.launcher.run_study(study_path, study_version)
+        job_uuid: UUID = self.launchers[launcher].run_study(
```
check if launcher exists and throw a custom exception

pl-buiquang commented on antarest/storage/repository/antares_io/exporter/export_file.py:
```diff
@@ -29,3 +30,34 @@ def export_file(self, path_study: Path, outputs: bool = True) -> BytesIO:
+        outputs: bool = False,
+    ) -> None:
+        ignore_patterns = (
+            shutil.ignore_patterns("output") if not outputs else None
```
this is not going to go well if there is an area named output..., should ignore only the strict path "path_study / output"

pl-buiquang commented on requirements.txt:
```diff
@@ -1,3 +1,5 @@
+-r antares-launcher/requirements.txt
+flask==1.1.2
```
you can remove this

pl-buiquang commented on resources/application.yaml:
```diff
@@ -24,6 +24,27 @@ launcher:
+      - "710"
+      - "720"
+      - "800"
+
```
you can comment all this (the resources/application.yaml is the default config for local mode and should work as is)

Hyralc committed:
varous fixes

make launchers optional
Fix output importer when the output directory does not exist
fix tests

Hyralc committed:
various changes

Hyralc committed:
Partially fix mypy

pl-buiquang committed:
Fixes

Signed-off-by: Paul Bui-Quang <paul.buiquang@rte-france.com>

pl-buiquang committed:
Fix dockerfile

Signed-off-by: Paul Bui-Quang <paul.buiquang@rte-france.com>

pl-buiquang committed:
More fixes..

Signed-off-by: Paul Bui-Quang <paul.buiquang@rte-france.com>

pl-buiquang committed:
Remove unused slurm config param

Signed-off-by: Paul Bui-Quang <paul.buiquang@rte-france.com>

pl-buiquang committed:
Fix callback

Signed-off-by: Paul Bui-Quang <paul.buiquang@rte-france.com>

pl-buiquang committed:
Fix test 'oups'

Signed-off-by: Paul Bui-Quang <paul.buiquang@rte-france.com>

Hyralc committed:
test run study

Hyralc committed:
test check state

Hyralc committed:
small changes

Hyralc committed:
optimize imports

pl-buiquang committed:
Merge branch 'dev' into Add_launcher_service

pl-buiquang committed:
Fix mypy

Signed-off-by: Paul Bui-Quang <paul.buiquang@rte-france.com>
